### PR TITLE
SG-15693 add change context mode

### PR DESCRIPTION
**Pull request by pscadding**
Adds a context change mode that means the open and save app options are replaced by a context change app.
In context change mode it contains the entity/context navigation windows, but no file select. Once a context has been picked you can press change context.

Tasks in the My tasks can be view can be double clicked to change context, when in context change mode.

**Refactoring of the Open, Save and Context change forms**

Previously, the open and save dialog derive from the same `FileFormBase` class, but that class makes no assumption about the UI, because the save and open dialogs have different UIs.
However this meant a bunch of duplicated code in the derived classes, because both UI's shared common functionality.
Instead of opting to make context change a "mode" of the open dialog, I decided to make it it's own class derived from the `FileFormBase` and to cut down on the duplication across all three classes I moved a bunch of methods over to the base class. However this means that the base class now has some expectation of what will be in the UI.
I haven't but I could make the ui elements required by the base class, properties which the deriving classes must implement, although the element names would be identical in all cases so I felt this was overkill. Or I could revert to duplicating up the code again across all three classes.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -37,43 +37,64 @@ def init_app(self):
...
-                # dark themed icon for engines that recognize this format
-                "icons": {
-                    "dark": {
-                        "png": os.path.join(
-                            os.path.dirname(__file__),
-                            "resources",
-                            "file_open_menu_icon.png",
-                        )
-                    }
+        if self.get_setting("context_change_only_mode"):
```
Comment (jfboismenu) [inline app.py:40]: I think a more usefull approach for clients would be to have the following setting and their defaults

```
show_file_open: true
show_file_save: true
shot_change_context: false
```

This way we allow clients to select what they want. What do you think?

File: python/tk_multi_workfiles/actions/context_change_action.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Copyright (c) 2015 Shotgun Software Inc.
```
Comment (jfboismenu) [inline python/tk_multi_workfiles/actions/context_change_action.py:1]:
```suggestion
# Copyright (c) 2020 Shotgun Software Inc.
```

File: python/tk_multi_workfiles/actions/context_change_action.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Action to create a new file.
```
Comment (jfboismenu) [inline python/tk_multi_workfiles/actions/context_change_action.py:12]:
```suggestion
Action to create a change context.
```

File: python/tk_multi_workfiles/actions/context_change_action.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        Constructor.
+
+        :param environment: A WorkArea instance containing the context we will switch to.
+        """
+        Action.__init__(self, "Change Context")
+        self._environment = environment
+
+    def execute(self, parent_ui):
+        """
+        Perform a new-scene operation initialized with the current context.
```
Comment (jfboismenu) [inline python/tk_multi_workfiles/actions/context_change_action.py:38]:
```suggestion
        Perform a context change operation.
```

File: python/tk_multi_workfiles/actions/context_change_action.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                self._app.log_exception("Unable to run folder creation!")
+
+            if not self._environment.context == self._app.context:
+                # Change context
+                FileAction.change_context(self._environment.context)
+
+        except Exception as e:
+            error_title = "Failed to complete '%s' action" % self.label
+            QtGui.QMessageBox.information(
+                parent_ui, "%s" % error_title, "%s:\n\n%s" % (error_title, e)
```
Comment (jfboismenu) [inline python/tk_multi_workfiles/actions/context_change_action.py:62]:
```suggestion
                parent_ui, error_title, "%s:\n\n%s" % (error_title, e)
```
No need for string interpolation if we only use error_title as is 😉

File: python/tk_multi_workfiles/work_files.py [hunk 6]
```diff
@@ -102,7 +102,17 @@ def show_file_open_dlg():
         handler = WorkFiles()
         from .file_open_form import FileOpenForm
 
-        handler._show_file_dlg("File Open", FileOpenForm)
+        handler._show_file_dlg("", FileOpenForm)
```
Comment (jfboismenu) [inline python/tk_multi_workfiles/work_files.py:105]: Is that a mistake? It doesn't say File Open anymore.

**Review by jfboismenu — CHANGES_REQUESTED** [review]
Nice job! This is much cleaner that I expected it would be. I like that we were able to refactor more code to the base class. What you did wasn't unreasonable. Each dialog have the same general layout, so it makes sense for some of the callbacks to be shared.

There's one likely bug in there however, I think File Open does not have a title anymore.

I assume you've carefully went through the remove code in file_open_form and file_save_form to make sure the callbacks were exactly the same?

Awesome work!

File: app.py (hunk 1 above)
Comment (pscadding) [reply app.py:40]: I don't really, mind, I guess it allows more flexibly. I'm not sure when I would have the open app and not the save app, I guess maybe in a project context

File: python/tk_multi_workfiles/actions/context_change_action.py (hunk 4 above)
Comment (pscadding) [reply python/tk_multi_workfiles/actions/context_change_action.py:38]: thats annoying I did a pass on this file, to make sure I'd renamed it. It was obviously a bad pass :(

File: python/tk_multi_workfiles/actions/context_change_action.py (hunk 5 above)
Comment (pscadding) [reply python/tk_multi_workfiles/actions/context_change_action.py:62]: sure! I think I just copied it from somewhere else in the code, so that should be changed as well, I'll have a look.

File: python/tk_multi_workfiles/work_files.py (hunk 6 above)
Comment (pscadding) [reply python/tk_multi_workfiles/work_files.py:105]: I just wanted to mess with people. I have no idea how that crept in, thanks!

**Comment by pscadding** [comment]
I've made a few small changes to get it to work with Mari, you can see the diff here:
https://github.com/shotgunsoftware/tk-multi-workfiles2/compare/ce04f21..9729fe6

File: python/tk_multi_workfiles/actions/context_change_action.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        Perform a context change operation.
+
+        :param parent_ui: Parent dialog executing this action.
+        """
+
+        try:
+            # create folders and validate that we can save using the work template:
+            try:
+                # create folders if needed:
+                FileAction.create_folders(self._environment.context)
```
Comment (jfboismenu) [inline python/tk_multi_workfiles/actions/context_change_action.py:47]: Why the change from `create_folders_if_needed` to `create_folders`? I'm looking at the code and I understand the intent, but I don't understand it in this context (no pun intended 😜).

File: app.py [hunk 8]
```diff
@@ -32,48 +32,65 @@ def init_app(self):
             )
             return
 
-        if self.engine.name == "tk-mari":
```
Comment (jfboismenu) [inline app.py:35]: Sounds like asking to actually test in mari paid off. 😃

File: hooks/scene_operation_tk-mari.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+        :returns:               Depends on operation:
+                                'current_path' - Return the current scene
+                                                 file path as a String
+                                'reset'        - True if scene was reset to an empty
+                                                 state, otherwise False
+                                all others     - None
+        """
+
+        # Mari doesn't have any scene operations.
```
Comment (jfboismenu) [inline hooks/scene_operation_tk-mari.py:66]: Maybe we could elaborate a bit more here, explaining that the Mari workflow is based on context switching instead of opening scene files?

**Review by jfboismenu — COMMENTED** [review]
I'm unsure about some of your changes. Could you clarify the comments and explain why the change from `create_folders_if_needed` to `create_folders`?

File: python/tk_multi_workfiles/actions/context_change_action.py (hunk 7 above)
Comment (pscadding) [reply python/tk_multi_workfiles/actions/context_change_action.py:47]: If you don't actually have a template then the `create_folders_if_needed` method will error since it expects a template. With Mari you won't have a template, and if you are changing context even if you set it up in Maya, it doesn't necessarily make sense to expect a template either.

As far as I can gather from the `create_folders_if_needed` method, it is there to try and avoid folder creation if possible based on resolving templates. Since we can't rely on that, creating the folders without the check seems like the better option.

I had initially started to modify that method to check if there was a template, but I thought it probably just makes sense to call the `create_folders` instead.

Comment (pscadding) [reply python/tk_multi_workfiles/actions/context_change_action.py:47]: I guess going back to using `create_folders_if_needed` and then modifying it, would mean that if you say were using it in Maya where a template did exist, you could avoid the folder creation overhead, potentially. 
I'm happy to go with either approach.

**Review by jfboismenu — APPROVED** [review]
Everything is awesome! Thank you!
